### ISSUE TITLE
sshd_use_aproved_ciphers always fail even with the remediation applied

### ISSUE THREAD

**Issue by ptitoliv** [OpenSCAP/scap-security-guide#2155]
The rule checking sshd_use_aproved_ciphers always fail whereas the remediation has been correctly applied. The Ciphers list is correctly set in the /etc/ssh/sshd_config but when executing openscap eval, the rule is always in fail mode.

**Comment by mpreisler** [comment]
We need versions of SSG and OpenSCAP that you used. IIRC we fixed this quite recently.

**Comment by ptitoliv** [comment]
First, I was using an old version (0.1.30) and then I tried with the SSG master branch head fetched today. For OpenSCAP, I use the official RPM 1.2.10-2.

**Comment by redhatrises** [comment]
Can you provide `grep Ciphers /etc/ssh/sshd_config` output as well as if ssh is running and what OS you are using?

**Comment by ptitoliv** [comment]
Ciphers line is : Ciphers aes128-ctr, aes192-ctr,aes256-ctr
Running on CentOS 7.3.1611

**Comment by redhatrises** [comment]
@ptitoliv this test will always fail as you are not using a certified FIPS operating system.

**Comment by ptitoliv** [comment]
@redhatrises : Ok I understand the idea. However, wouldn't it be nice/possible to have a specific test for the SSG CentOS version which would test that we have strong Ciphers without considering the FIPS certification criteria ? I think it can be interesting for the CentOS users to verify the ciphers strength ? 

I guess it will be the same for the HMAC check ?

**Comment by shawndwells** [comment]
In theory yes. But that would have to be a non-US Gov related profile. Formal baselines (NIST, USGCB, STIG, CJIS, etc, all require FIPS). And thinking this through a bit more, non-Gov profiles like PCI and CSP also require FIPS.

**Comment by redhatrises** [comment]
> I guess it will be the same for the HMAC check ?

That is correct.

> Ok I understand the idea. However, wouldn't it be nice/possible to have a specific test for the SSG CentOS version which would test that we have strong Ciphers without considering the FIPS certification criteria ?

You need to modify the profile to use the `sshd_use_stron_ciphers` and `sshd_use_strong_macs` checks.

**Comment by ptitoliv** [comment]
Ok I got it. I just did a simulation replacing the sshd_use_approved_ciphers bu sshd_use_strong_ciphers and it works. The check is OK. Sorry for not having seen these others checkers in the source code.

However I have several question about CentOS and STIG : 
 - Wouldn't it be better if the FIPS check would be tagged as non applicable for CentOS in stead of being marked failed ?
 - Wouldn't it be nice to have a SSG for CentOS with non gov profiles derivated from the gov profiles excluding certifications things like FIPS rules I think it we be a good added value for CentOS users.

**Comment by redhatrises** [comment]
> Wouldn't it be better if the FIPS check would be tagged as non applicable for CentOS in stead of being marked failed ?

In this case, I don't believe that you can return "not applicable" for results enumeration. Plus, technically, it is a fail of the check as CentOS cannot meet the *legal* requirements.

> Wouldn't it be nice to have a SSG for CentOS with non gov profiles derivated from the gov profiles excluding certifications things like FIPS rules I think it we be a good added value for CentOS users.

Which non-Gov profiles would you recommend? See Shawn's note above about PCI and CSP also requiring FIPS.

**Comment by ptitoliv** [comment]
> In this case, I don't believe that you can return "not applicable" for results enumeration. Plus, technically, it is a fail of the check as CentOS cannot meet the legal requirements.

I see. I'm not yet a SCAP expert so I thought it was easy to make a check not applicable considering criteria. But referring to your explaination, it makes sense. 

> Which non-Gov profiles would you recommend? See Shawn's note above about PCI and CSP also requiring FIPS.

Well, actually, I was thinking about the following use case : I'm a CentOS user for some projects because it's perfect for my needs. I also need to have a compliancy with a security guide like STIG, CIS or whatever. 

However, my need is to know how secure my installation is without the need of a complete FIPS certification. So, what interests me is to know how i am compliant against a known guide like STIG which is the most interesting to me. However, if I use the SSG STIG profile, I will have some fails especially the one referring to FIPS rules.

So I thought it maybe would interesting to create an "unofficial" / "derivative" profil forked from the current STIG profile but removing the FIPS rules and replacing them by the matching one without the certifited stuff (Like sshd_use_approved_ciphers vs sshd_use_strong_ciphers). 

In a more detailed example, with the current STIG profile, the SSH cipher check will fail on CentOS because of the lack of FIPS certification. My idea would be to create a STIG alternative where we would replace the FIPS rules for the cipher by the non FIPS rule.

It would allow CentOS users using SSG to have a better overview considering a known guide. But of course this is just an idea.

**Comment by mpreisler** [comment]
@ptitoliv what about running the official profiles and being aware of why some rules fail on OSes without FIPS cert? Duplicating every profile would bloat the content quite a lot and would make UX horrible in SCAP Workbench and other GUI tools.

**Comment by mpreisler** [comment]
@ptitoliv You can even create these profiles yourself using workbench. And take the FIPS rules out of your profiles. Check out http://static.open-scap.org/scap-workbench-1.1

**Comment by ptitoliv** [comment]
@mpreisler, to answer to your two questions : This is what I actually do for some checks : I put a justification for checks i'm not compliant with and I also disable using scap-workbench checks I know i will never be compliant with.

So of course I can remove the FIPS rules using scap-workbench but what I find a little bit annoying is not having a replacing rule for CentOS that will have a check pass which would avoid me a justification.

As I told you in the beginning of my answer, I customized my profile and i maintain it against the SSG upgrades. But I was thinking it could be nice to have the choice especially for the CentOS user who will never have the FIPS certification.

I fully understand that the copy work would be horrible (lots of duplication useless work). 

Would it be possible to put the two rules into the STIG profile and let the FIPS one enabled by default and the Strong Cipher disabled (I still use my sshd cipher example) and we would use scap-workbench to do the change. I guess this is not possible because of STIG legal compliancy ?. 

Maybe i'm fully wrong with my point of view. It's just I think it would be a nice feature for the SSG CentOS version.

**Comment by shawndwells** [comment]
On 7/13/17 5:23 PM, Olivier Bonhomme wrote:
> <snip>
>
> Would it be possible to put the two rules into the STIG profile and
> let the FIPS one enabled by default and the Strong Cipher disabled (I
> still use my sshd cipher example) and we would use scap-workbench to
> do the change. I guess this is not possible because of STIG legal
> compliancy ?.
>
> </snip>

Configuration checks can be included in the larger body of compliance
content, but not selected in specific profiles. This would let people
create customized profiles with those rules without impacting known
baselines like STIG, PCI, etc.

There are significant ethical issues with this though. People running
CentOS should absolutely fail every single cryptography check. It is
*illegal* to run CentOS in the US Government because lack of FIPS.
Beyond lacking crypto evaluations, CentOS contains roughly 33% different
binaries than RHEL. Creating any sense that CentOS uses "strong ciphers"
is extremely misleading.... even a rule such as "Use Strong Ciphers"
should fail on CentOS (but pass on Ubuntu, SuSE, RHEL, which have
evaluated crypto).

**Comment by ptitoliv** [comment]
> Configuration checks can be included in the larger body of compliance
> content, but not selected in specific profiles. This would let people
> create customized profiles with those rules without impacting known
> baselines like STIG, PCI, etc.

Do you mean something like this in the generated XCCDF ? 

```
<ns0:select idref="sshd_use_approved_ciphers" selected="true">
<ns0:select idref="sshd_use_strong_ciphers" selected="false">
```
If yes, I definitely vote for that solution.

> There are significant ethical issues with this though. People running
> CentOS should absolutely fail every single cryptography check. It is
> *illegal* to run CentOS in the US Government because lack of FIPS.

I fully understand that point of view. However, I think it could be nice if SSG could be used also into a community usage where this strong requirement is not applicable.

> Beyond lacking crypto evaluations, CentOS contains roughly 33% different
> binaries than RHEL. Creating any sense that CentOS uses "strong ciphers"
> is extremely misleading.... even a rule such as "Use Strong Ciphers"
> should fail on CentOS (but pass on Ubuntu, SuSE, RHEL, which have
> evaluated crypto).

When I do a project using CentOS, I am perfectly aware that I won't benefit of any crypto certification. This is not what I am looking for when I use CentOS. However, what interests me is to know if my SSH configuration doesn't use any weak ciphers and nothing more. That's why I proposed that idea of an SSG profile dedicated for community distributions (It could be also applied to Debian) with alternatives rules in order to be able to have a good compliancy against a reference guide like STIG.

**Comment by redhatrises** [comment]
Closing as creating a customized profile is what should happen in this case. Using invalid STIG rules in an official STIG profile should not be implemented for any reason as the rules do not meet STIG compliance.

**Comment by ptitoliv** [comment]
Hello @redhatrises, @shawndwells and @mpreisler.

Thank you for all your interesting answers. Actually there was a misunderstanding from me using scap-workbench and the last version of the SSG I pulled from git. I thought that the non-FIPS rules weren't integrated in the SSG XCCDF file because they were not enabled into the provided profiles.  But actually the rules are available so I can do what @redhatrises  told me and creating my custom profile with the content provided by the SSG without any code modification.

So, sorry for the noise about that misunderstanding coming from me :( But thanks to that, I learned very interesting things about FIPS certification and rules.

Hope my next contributions will be more relevant since I'm very interested in doing work with OpenSCAP for community distributions.

Thank you.
